File: app/src/main/java/com/elementary/tasks/navigation/settings/LocationSettingsFragment.java
package com.elementary.tasks.navigation.settings;

import android.app.AlertDialog;
import android.os.Bundle;
import androidx.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.SeekBar;

import com.elementary.tasks.R;
import com.elementary.tasks.core.utils.Constants;
import com.elementary.tasks.core.utils.Dialogues;
import com.elementary.tasks.core.utils.Prefs;
import com.elementary.tasks.core.utils.ThemeUtil;
import com.elementary.tasks.databinding.DialogTrackingSettingsLayoutBinding;
import com.elementary.tasks.databinding.FragmentSettingsLocationBinding;
import com.elementary.tasks.navigation.settings.location.MapStyleFragment;
import com.elementary.tasks.navigation.settings.location.MarkerStyleFragment;

import java.util.Locale;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public class LocationSettingsFragment extends BaseSettingsFragment {

    private int mItemSelect;

    private FragmentSettingsLocationBinding binding;
    private View.OnClickListener mRadiusClick = view -> showRadiusPickerDialog();
    private View.OnClickListener mNotificationClick = view -> changeNotificationPrefs();
    private View.OnClickListener mMapTypeClick = view -> showMapTypeDialog();
    private View.OnClickListener mStyleClick = view -> replaceFragment(new MarkerStyleFragment(), getString(R.string.style_of_marker));
    private View.OnClickListener mTrackClick = view -> showTrackerOptionsDialog();

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentSettingsLocationBinding.inflate(inflater, container, false);
        initMapTypePrefs();
        initMarkerStylePrefs();
        binding.trackerPrefs.setOnClickListener(mTrackClick);
        binding.notificationOptionPrefs.setOnClickListener(mNotificationClick);
        binding.notificationOptionPrefs.setChecked(getPrefs().isDistanceNotificationEnabled());
        initRadiusPrefs();
        return binding.getRoot();
    }

    private void initMapStylePrefs() {
        binding.mapStylePrefs.setOnClickListener(v -> openMapStylesFragment());
        binding.mapStylePrefs.setDetailText(getString(ThemeUtil.getInstance(getContext()).getStyleName()));
        binding.mapStylePrefs.setViewResource(ThemeUtil.getInstance(getContext()).getMapStylePreview());
        binding.mapStylePrefs.setEnabled(Prefs.getInstance(getContext()).getMapType() == 3);
    }

    private void openMapStylesFragment() {
        if (getCallback() != null) {
            getCallback().replaceFragment(MapStyleFragment.newInstance(), getString(R.string.map_style));
        }
    }

    private void initMarkerStylePrefs() {
        binding.markerStylePrefs.setOnClickListener(mStyleClick);
        showMarkerStyle();
    }

    private void showMarkerStyle() {
        binding.markerStylePrefs.setViewResource(ThemeUtil.getInstance(getContext()).getMarkerStyle());
    }

    private void initMapTypePrefs() {
        binding.mapTypePrefs.setOnClickListener(mMapTypeClick);
        showMapType();
    }

    private void showMapType() {
        String[] types = getResources().getStringArray(R.array.map_types);
        binding.mapTypePrefs.setDetailText(types[getPosition(getPrefs().getMapType())]);
    }

    private void initRadiusPrefs() {
        binding.radiusPrefs.setOnClickListener(mRadiusClick);
        showRadius();
    }

    @Override
    public void onResume() {
        super.onResume();
        showMarkerStyle();
        initMapStylePrefs();
        if (getCallback() != null) {
            getCallback().onTitleChange(getString(R.string.location));
            getCallback().onFragmentSelect(this);
        }
    }

    private void showTrackerOptionsDialog(){
        AlertDialog.Builder builder = Dialogues.getDialog(getContext());
        builder.setTitle(R.string.tracking_settings);
        DialogTrackingSettingsLayoutBinding b = DialogTrackingSettingsLayoutBinding.inflate(LayoutInflater.from(getContext()));
        b.distanceBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                b.distanceTitle.setText(String.format(Locale.getDefault(), getString(R.string.x_meters), String.valueOf(progress + 1)));
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {

            }
        });
        int distance = getPrefs().getTrackDistance() - 1;
        b.distanceBar.setProgress(distance);
        b.distanceTitle.setText(String.format(Locale.getDefault(), getString(R.string.x_meters), String.valueOf(distance + 1)));
        b.timeBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                b.timeTitle.setText(String.format(Locale.getDefault(), getString(R.string.x_seconds), String.valueOf(progress + 1)));
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {

            }
        });
        int time = getPrefs().getTrackTime() - 1;
        b.timeBar.setProgress(time);
        b.timeTitle.setText(String.format(Locale.getDefault(), getString(R.string.x_seconds), String.valueOf(time + 1)));
        builder.setView(b.getRoot());
        builder.setPositiveButton(R.string.ok, (dialogInterface, i) -> {
            getPrefs().setTrackDistance(b.distanceBar.getProgress() + 1);
            getPrefs().setTrackTime(b.timeBar.getProgress() + 1);
        });
        builder.setNegativeButton(R.string.cancel, (dialog, which) -> dialog.dismiss());
        builder.create().show();
    }

    private void showMapTypeDialog() {
        AlertDialog.Builder builder = Dialogues.getDialog(getContext());
        builder.setCancelable(true);
        builder.setTitle(getString(R.string.map_type));
        String[] types = new String[]{
                getString(R.string.normal),
                getString(R.string.satellite),
                getString(R.string.terrain),
                getString(R.string.hybrid)
        };
        ArrayAdapter<String> adapter = new ArrayAdapter<>(getContext(),
                android.R.layout.simple_list_item_single_choice, types);
        int type = getPrefs().getMapType();
        mItemSelect = getPosition(type);
        builder.setSingleChoiceItems(adapter, mItemSelect, (dialog, which) -> mItemSelect = which);
        builder.setPositiveButton(getString(R.string.ok), (dialogInterface, i) -> {
            getPrefs().setMapType(mItemSelect + 1);
            showMapType();
            initMapStylePrefs();
            dialogInterface.dismiss();
        });
        AlertDialog dialog = builder.create();
        dialog.setOnCancelListener(dialogInterface -> mItemSelect = 0);
        dialog.setOnDismissListener(dialogInterface -> mItemSelect = 0);
        dialog.show();
    }

    private int getPosition(int type) {
        int mItemSelect;
        if (type == Constants.MAP_SATELLITE){
            mItemSelect = 1;
        } else if (type == Constants.MAP_TERRAIN){
            mItemSelect = 2;
        } else if (type == Constants.MAP_HYBRID){
            mItemSelect = 3;
        } else {
            mItemSelect = 0;
        }
        return mItemSelect;
    }

    private void changeNotificationPrefs() {
        boolean isChecked = binding.notificationOptionPrefs.isChecked();
        binding.notificationOptionPrefs.setChecked(!isChecked);
        getPrefs().setDistanceNotificationEnabled(!isChecked);
    }

    private void showRadius() {
        binding.radiusPrefs.setDetailText(String.format(Locale.getDefault(), getString(R.string.radius_x_meters),
                String.valueOf(getPrefs().getRadius())));
    }

    private void showRadiusPickerDialog(){
        int radius = getPrefs().getRadius();
        Dialogues.showRadiusDialog(getContext(), radius, new Dialogues.OnValueSelectedListener<Integer>() {
            @Override
            public void onSelected(Integer integer) {
                getPrefs().setRadius(integer);
                showRadius();
            }

            @Override
            public String getTitle(Integer integer) {
                return String.format(Locale.getDefault(), getString(R.string.radius_x_meters),
                        String.valueOf(integer));
            }
        });
    }
}

File: app/src/main/java/com/elementary/tasks/navigation/settings/LocationSettingsFragment.kt
package com.elementary.tasks.navigation.settings

import android.os.Bundle
import android.view.View
import android.widget.ArrayAdapter
import android.widget.SeekBar
import com.elementary.tasks.R
import com.elementary.tasks.core.utils.Constants
import com.elementary.tasks.core.utils.Dialogues
import com.elementary.tasks.navigation.settings.location.MapStyleFragment
import com.elementary.tasks.navigation.settings.location.MarkerStyleFragment
import kotlinx.android.synthetic.main.dialog_tracking_settings_layout.view.*
import kotlinx.android.synthetic.main.fragment_settings_location.*
import java.util.*

/**
 * Copyright 2016 Nazar Suhovich
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
class LocationSettingsFragment : BaseSettingsFragment() {

    private var mItemSelect: Int = 0

    override fun layoutRes(): Int = R.layout.fragment_settings_location

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initMapTypePrefs()
        initMarkerStylePrefs()
        trackerPrefs.setOnClickListener { showTrackerOptionsDialog() }
        notificationOptionPrefs.setOnClickListener { changeNotificationPrefs() }
        notificationOptionPrefs.isChecked = prefs.isDistanceNotificationEnabled
        initRadiusPrefs()
    }

    private fun initMapStylePrefs() {
        mapStylePrefs.setOnClickListener { openMapStylesFragment() }
        mapStylePrefs.setDetailText(getString(themeUtil.styleName))
        mapStylePrefs.setViewResource(themeUtil.mapStylePreview)
        mapStylePrefs.isEnabled = prefs.mapType == 3
    }

    private fun openMapStylesFragment() {
        if (callback != null) {
            callback?.replaceFragment(MapStyleFragment.newInstance(), getString(R.string.map_style))
        }
    }

    private fun initMarkerStylePrefs() {
        markerStylePrefs.setOnClickListener { replaceFragment(MarkerStyleFragment(), getString(R.string.style_of_marker)) }
        showMarkerStyle()
    }

    private fun showMarkerStyle() {
        markerStylePrefs.setViewResource(themeUtil.markerStyle)
    }

    private fun initMapTypePrefs() {
        mapTypePrefs.setOnClickListener { showMapTypeDialog() }
        showMapType()
    }

    private fun showMapType() {
        val types = resources.getStringArray(R.array.map_types)
        mapTypePrefs.setDetailText(types[getPosition(prefs.mapType)])
    }

    private fun initRadiusPrefs() {
        radiusPrefs.setOnClickListener { showRadiusPickerDialog() }
        showRadius()
    }

    override fun onResume() {
        super.onResume()
        showMarkerStyle()
        initMapStylePrefs()
        if (callback != null) {
            callback?.onTitleChange(getString(R.string.location))
            callback?.onFragmentSelect(this)
        }
    }

    private fun showTrackerOptionsDialog() {
        val builder = dialogues.getDialog(context!!)
        builder.setTitle(R.string.tracking_settings)
        val b = layoutInflater.inflate(R.layout.dialog_tracking_settings_layout, null)
        b.distanceBar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                b.distanceTitle.text = String.format(Locale.getDefault(), getString(R.string.x_meters), (progress + 1).toString())
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {

            }

            override fun onStopTrackingTouch(seekBar: SeekBar) {

            }
        })
        val distance = prefs.trackDistance - 1
        b.distanceBar.progress = distance
        b.distanceTitle.text = String.format(Locale.getDefault(), getString(R.string.x_meters), (distance + 1).toString())
        b.timeBar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                b.timeTitle.text = String.format(Locale.getDefault(), getString(R.string.x_seconds), (progress + 1).toString())
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {

            }

            override fun onStopTrackingTouch(seekBar: SeekBar) {

            }
        })
        val time = prefs.trackTime - 1
        b.timeBar.progress = time
        b.timeTitle.text = String.format(Locale.getDefault(), getString(R.string.x_seconds), (time + 1).toString())
        builder.setView(b)
        builder.setPositiveButton(R.string.ok) { _, _ ->
            prefs.trackDistance = b.distanceBar.progress + 1
            prefs.trackTime = b.timeBar.progress + 1
        }
        builder.setNegativeButton(R.string.cancel) { dialog, _ -> dialog.dismiss() }
        builder.create().show()
    }

    private fun showMapTypeDialog() {
        val builder = dialogues.getDialog(context!!)
        builder.setCancelable(true)
        builder.setTitle(getString(R.string.map_type))
        val types = arrayOf(getString(R.string.normal), getString(R.string.satellite), getString(R.string.terrain), getString(R.string.hybrid))
        val adapter = ArrayAdapter(context!!, android.R.layout.simple_list_item_single_choice, types)
        val type = prefs.mapType
        mItemSelect = getPosition(type)
        builder.setSingleChoiceItems(adapter, mItemSelect) { _, which -> mItemSelect = which }
        builder.setPositiveButton(getString(R.string.ok)) { dialogInterface, _ ->
            prefs.mapType = mItemSelect + 1
            showMapType()
            initMapStylePrefs()
            dialogInterface.dismiss()
        }
        val dialog = builder.create()
        dialog.setOnCancelListener { mItemSelect = 0 }
        dialog.setOnDismissListener { mItemSelect = 0 }
        dialog.show()
    }

    private fun getPosition(type: Int): Int {
        return when (type) {
            Constants.MAP_SATELLITE -> 1
            Constants.MAP_TERRAIN -> 2
            Constants.MAP_HYBRID -> 3
            else -> 0
        }
    }

    private fun changeNotificationPrefs() {
        val isChecked = notificationOptionPrefs.isChecked
        notificationOptionPrefs.isChecked = !isChecked
        prefs.isDistanceNotificationEnabled = !isChecked
    }

    private fun showRadius() {
        radiusPrefs.setDetailText(String.format(Locale.getDefault(), getString(R.string.radius_x_meters),
                prefs.radius.toString()))
    }

    private fun showRadiusPickerDialog() {
        val radius = prefs.radius
        dialogues.showRadiusDialog(context!!, radius, object : Dialogues.OnValueSelectedListener<Int> {
            override fun onSelected(integer: Int) {
                prefs.radius = integer
                showRadius()
            }

            override fun getTitle(integer: Int): String {
                return String.format(Locale.getDefault(), getString(R.string.radius_x_meters),
                        integer.toString())
            }
        })
    }
}
